### PR TITLE
Allows compiling with clang + Visual Studio 2019

### PR DESCRIPTION
**Pull request by francois-keith**

### CODE REVIEW

File: unittest/CMakeLists.txt [hunk 1]
```diff
@@ -118,9 +126,11 @@ ADD_PINOCCHIO_UNIT_TEST(liegroups)
 ADD_PINOCCHIO_UNIT_TEST(regressor)
 ADD_PINOCCHIO_UNIT_TEST(version)
 ADD_PINOCCHIO_UNIT_TEST(copy)
-MAKE_DIRECTORY("${CMAKE_CURRENT_BINARY_DIR}/serialization-data")
-ADD_PINOCCHIO_UNIT_TEST(serialization)
-ADD_TEST_CFLAGS(serialization '-DTEST_SERIALIZATION_FOLDER=\\\"${CMAKE_CURRENT_BINARY_DIR}/serialization-data\\\"')
+if(NOT WIN32)
```
Comment (jcarpent) [inline unittest/CMakeLists.txt:129]: @francois-keith Do you know why this test is not passing?

**Comment by jcarpent** [comment]
By the way, do you have tried to compile Python bindings too? If yes, do you have encountered  issues when running them?

I know that there is a small mismatch between the way to generate Python bindings on Win32, that may require some adjustements in EigenPy.

File: unittest/CMakeLists.txt (hunk 1 above)
Comment (francois-keith) [reply unittest/CMakeLists.txt]: Yes, I just figured it out. I used to have a link issue, but it was due to a faulty boost installation. Once corrected, I had an input stream error in one of the unit tests, solved by 02e4c3d

**Comment by francois-keith** [comment]
I'll have look for the ci-build and the python bindings (I only tried compiling the c++ bindings).

**Comment by francois-keith** [comment]
Regarding the python build, it fails with the following error :

```
lld-link : error : undefined symbol: EIGENPY_ARRAY_API
>> referenced by pinocchio_pywrap.dir\Release\module.obj:(public: static void * __cdecl eigenpy::EigenFromPy<class Eigen::Matrix<double, 6, 6, 0, 6, 6>>::convertible(struct tagPyArrayObject *))
>> ... (+ other references in module.obj)
```

Note that eigenpy.lib is correctly included (otherwise there would be other link errors).

This value is declared [there:](https://github.com/stack-of-tasks/eigenpy/blob/master/include/eigenpy/numpy.hpp#L12)

```
#ifndef PY_ARRAY_UNIQUE_SYMBOL
  #define PY_ARRAY_UNIQUE_SYMBOL EIGENPY_ARRAY_API
#endif
```

But I don't quite get where EIGENPY_ARRAY_API is defined, so I cannot really correct this issue.

**Comment by jcarpent** [comment]
Hi @francois-keith,

I've just published a new release of EigenPy, with some fixes with respect to what was a bug on Windows. Could you try it and give me some feedback?

Thanks in advance,
Best,

Justin

**Comment by francois-keith** [comment]
@jcarpent Do you mind if I rebase on stack-of-tasks:devel and push-force on this branch ?

**Comment by jcarpent** [comment]
This is fine for me. Please proceed. In the meantime, I'm changing the CMake file to not rely on pkg-config by default and much more used the standard CMake approach. @francois-keith Is it fine for you too?

**Comment by francois-keith** [comment]
More than fine ! 😃

**Comment by francois-keith** [comment]
⚠️ The positive result above cannot be trusted: the cmake generation fails on win32, but this is considered as a success.

**Comment by jcarpent** [comment]
I've tried to compile Pinocchio with Visual Studio 2019, but it totally fails. @francois-keith How much Visual can handle templated frameworks? Which compiler do you use?

**Comment by jcarpent** [comment]
@sbarthelemy Which compiler do you use on Windows?

**Comment by francois-keith** [comment]
I used clang to compile pinocchio, otherwise there was indeed some template issues, even with vs19

To compile with clang you need to add the option -T"ClangCl" in your cmake command line.

**Comment by jcarpent** [comment]
This is now working better! Thanks @francois-keith!

**Comment by jcarpent** [comment]
@francois-keith I'm almost done on my side. Everything seems to work well on my local machine.
The only issue I got is that I'm not able to make "ClangCl" working on the CI.

I was wondering how to setup it, but I cannot find any help on the web.
If you have any clue ...

**Comment by jcarpent** [comment]
@francois-keith The issue is now fixed. It was due to a bug in cmake provided by conda on the conda-forge channel.
I'm now facing another issue, which concerns the apparition of some `near` and `far` names variables inside urdfdom (https://github.com/ros/urdfdom_headers/blob/a15d906ff16a7fcbf037687b9c63b946c0cc04a1/urdf_sensor/include/urdf_sensor/sensor.h#L93), which seems to be reserved by Windows. 

So, I've provided a fix whihc consists of those lines:
```
#ifdef _WIN32
  #include <windows.h>
  #ifdef far
    #undef far
  #endif
  #if def near
    #undef near
  #endif
#endif
```

@francois-keith Do you see other alternative? @traversaro Have you encountered such issue in the past with urdfdom?

**Comment by francois-keith** [comment]
@jcarpent I think this would work

**Comment by traversaro** [comment]
>  @traversaro Have you encountered such issue in the past with urdfdom?

Not with the `urdfdom`, even if `near`/`far` collision when including `windows.h` are a classic. However, which file is including `windows.h` ? In general including  `windows.h`  in public headers is something that should avoided as much as possible, exactly to avoid  this kind of problems.

**Comment by jcarpent** [comment]
This include is located inside the unsupported Eigen Tensor module, unfortunately.

**Comment by traversaro** [comment]
> 
> 
> This include is located inside the unsupported Eigen Tensor module, unfortunately.

Ah, I see, the related upstream issue is https://gitlab.com/libeigen/eigen/-/issues/1801 . For reference, it was added in https://gitlab.com/libeigen/eigen/-/commit/4deafd35b75cde9c9d40360a37c364594fd8161a .

**Comment by jmirabel** [comment]
For reference: #1224 should be addressed here.

**Comment by jcarpent** [comment]
This PR is not yet finished. I will handle #1224 and we need to update the CMake module to correctly handle unit testings.

**Comment by jcarpent** [comment]
Good news, everything is now working fine after some fixing in CMake and the related submodule.

**Pull request merged**